addons/crm_livechat/models/chatbot_script_step.py

```diff
@@ -62,4 +70,39 @@ def _process_step_create_lead(self, discuss_channel):
         create_values.update(self._chatbot_crm_prepare_lead_values(
             discuss_channel, customer_values['description']))
 
-        self.env['crm.lead'].create(create_values)
+        return self.env['crm.lead'].create(create_values)
+
+    def _process_step_create_lead_and_forward(self, discuss_channel, posted_message):
+        lead = self._process_step_create_lead(discuss_channel)
```

Comment:
The lines below are almost identical to what is already implemented in `_process_step_forward_operator`. However, your implementation does not rely on ` im_livechat_channel@_get_operator` and, as a result, does not follow the assignation rules. An operator will be assigned regardless of busyness, language, country, ongoing calls, number of session limits, etc. All assignations must go through `_get_operator`. This is the only way to ensure that all rules in place are respected.

In essence, the code should be as simple as:

```py
self._process_step_create_lead(channel)
self._process_step_forward_operator(channel)
```

We should only override `_compute_available_operators` to add the conditions related to `crm_team_member_ids` (I think `_get_operator` will do the rest).

Reply:
Isn't the goal here to have the step attempt to assign the specifically found sales person in priority?

How is changing `_compute_available_operators` going to take into account which sales person must be selected?

Reply:
There are two different things here:

1.

```py
                    if not member.assignment_optout
                    and member._get_assignment_quota() > 0
```

This condition should be in `_compute_available_operator_ids`. It's clearly a condition that removes some operators from the list of available operators, so we should only rely on the `available_operator_ids` fields to do it.

2.

The logic to forward the operator is clearly duplicated and does not rely on `_get_operator` nor on `_available_operator_ids` which will totally bypass the assignation logic (max_sessions, lang, ...). We could perfectly call `_get_operator` giving the operators to chose from (i.e. the crm team members). That would:
- prevent duplicated logic.
- ensure operator selection is done consistently according to the rules that are already in place.

Conceptually, this new step is just the combination of two existing steps. The fact that a salesperson should be assigned specifically is just a detail.

Reply:
I think what we should do is make `_process_step_forward_operator` take an optional operator as param then, so we can properly forward the assigned person.

We do need to check the availability before assignation though, so we will need to extract the logic out of `_available_operator_ids` (probably on a user method then, optionally taking the livechat channel).

I'm still not sure if the sales person must be part of the livechat channel and be listed in `_available_operator_ids`. I'm thinking it should not be there, as any other forward/step should not take the sales person into account for assignment.
I'll ask functionally what we actually want.

Reply:
> I think what we should do is make _process_step_forward_operator take an optional operator as param then, so > we can properly forward the assigned person.

I feel like it should be done the other way around: `_process_step_forward_operator` might take an optional array of operators to chose from (i.e. available operators from the crm team), and chose the best according to assignation rules? Then we can assign this person to the lead, don't you think?

Reply:
I’ve created a workaround with this in mind. Could you please take a look and let me know if this approach could work? It’s just a rough draft and will need refinements.